### DISCUSSION TITLE
[Book 1, Chapter 4.2] Z axis orientation

### DISCUSSION THREAD

**Discussion by dimitry-ishenko** [RayTracing/raytracing.github.io#1386] [General]
From chapter 4.2 of book 1:

> For simplicity we'll start with the camera center at (0,0,0). We'll also have the y-axis go up, the x-axis to the right, and **the negative z-axis pointing in the viewing direction.** (This is commonly referred to as right-handed coordinates.)
>
> <img src="https://github.com/RayTracing/raytracing.github.io/assets/2858291/f1468176-0593-44fd-85dc-de0bcfa92699)" width=50%>

It is my understanding that arrows on axes indicate positive direction. So, the above paragraph states that negative Z points towards the viewport, but the image shows the opposite -- the arrow (aka positive Z) is pointing towards the viewport.

Did I miss something?

**Comment by rupsis** [comment]
The coordinates here just act as a reference. _Typically_ we associate axis illustrations (pointing arrows) with positive direction, but that's not the case here. Not wrong, but potentially confusing. 

 I personally think we should always label the axes. @hollasch  do you have any thoughts? 
 
 p.s, linking to [Right hand rule](https://en.wikipedia.org/wiki/Right-hand_rule) might be helpful.

**Comment by dimitry-ishenko** [comment]
It definitely confused me. I try to write my own code and when I glanced at the illustration above to make sure I go my axis directions right, I ended up with inverted Z axis.

Something like this might be less confusing:

<img src="https://github.com/RayTracing/raytracing.github.io/assets/2858291/e187157e-523d-42a2-9080-56ce44f8f193" width="50%">

**Comment by dimitry-ishenko** [comment]
This is slightly OT, but why did you chose right-handed coordinates? From what I understand both OpenGL and DirectX both use left-handed one.

**Comment by armansito** [comment]
I agree that the diagram is confusing. I understand the arrows as demonstrating the camera's basis vectors (up, right, look) and not the coordinate axes but this distinction could be illustrated better. Somewhat relatedly, I'm currently working on a similar diagram for the ray generation chapter of [the GPU book](https://github.com/RayTracing/gpu-tracing/blob/dev/book/MovingToTheGPU.html#L1622) where I intend to illustrate this clearly.

> This is slightly OT, but why did you chose right-handed coordinates? From what I understand both OpenGL and DirectX both use left-handed one.

I can't speak for the authors' original choice but I think this is a matter of preference. The convention varies between graphics APIs (IIRC OpenGL defines a ~~right~~ left handed NDC system, Direct3D and Metal are left-handed, Vulkan is right-handed but with the $+y$-axis pointing down s.t. the view direction is towards $+z$).

It's ultimately a matter of convention. An application that targets more than one of these APIs will typically pick its own convention and reconcile any mismatch when computing the transformations. FWIW, I picked a right-handed ($+y$-up) system for the GPU book as it is common to work in right-handed world and object-space coordinates.

**Comment by hollasch** [comment]
Arrows indicate vector direction, always positive. But these are not axes; they are three independent vectors used to construct scene rays. They are not orthonormal, and they are not even _orthogonal_. In practice, the scene ray will move from target pixel to target pixel (and _around_ target pixels for sampling). These are _not_ the $X$, $Y$, or $Z$ basis vectors.

**Reply by dimitry-ishenko** [reply]
> Arrows indicate vector direction, always positive. But these are not axes;

@hollasch that's not what the paragraph above the picture says (see my original post). Quote from the paragraph: _"We'll also have the y-axis go up, the x-axis to the right, and the **negative** z-axis pointing in the viewing direction."_

Figure 3 just below it seems to visually show what was said in the paragraph, but **positive** z-axis is pointing in the viewing direction.

> They are not orthonormal

Not sure what this has to do with anything. We are talking about axis directions and not basis vectors.

> and they are not even orthogonal.

They _are_ according to the paragraph above. Are they not?

If they aren't, then I am even more confused by that paragraph and picture... 😕

**Comment by ParthKohli** [comment]
I was also confused by the z-axis orientation: the code seems to directly disagree with the diagram. The following in 1.4.2 indicates that the viewport is in the negative direction:

```cpp
auto viewport_upper_left = camera_center
                             - vec3(0, 0, focal_length) - viewport_u/2 - viewport_v/2;
```

But if right-handed coordinates were actually being followed, then applying the right-hand rule suggests that the viewport is in the positive z-direction.

<img src="https://raytracing.github.io/images/fig-1.03-cam-geom.jpg" height="150px">

The above code listing should be modified to the following throughout the code: 

```cpp
auto viewport_upper_left = camera_center
                             + vec3(0, 0, focal_length) - viewport_u/2 - viewport_v/2;
```
